docs/current/guides/757394-use-services.md

```diff
@@ -0,0 +1,326 @@
+
+- Each container has its own network namespace and IP address
+- Each container has a unique, deterministic DNS address
+- Containers can expose ports and endpoints
+- Containers can bind other containers as services
+
+Service containers come with the following built-in features:
+
+- Service containers are started just-in-time, de-duplicated, and stopped when no longer needed
+- Service containers are health checked prior to running clients
+- Service containers are given an alias for the client container to use as its hostname
+
+## Using hostnames
+
+Containers run in a bridge network. Each container has its own IP address that other containers can reach. Here's a simple example:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="blm2lRNoYzE" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="bhx9i4LCcD9" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="fR2OAVbNUVH" />
+
+</TabItem>
+</Tabs>
+
+Containers never use IP addresses to reach each other directly. IP addresses are ephemeral, so doing so would nullify the cache. Instead, Dagger gives each container a unique but deterministic hostname, which doubles as a DNS address. Here's an example:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="H5Eb0Hs7JMd" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="2qLVfgdsnI6" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="Bpu7I8URtpg" />
+
+</TabItem>
+</Tabs>
+
+This hash value is derived from the same value that determines whether an operation is a cache hit in Buildkit: the vertex digest.
+
+To get a container's address, you wouldn't normally run the `hostname` command, because you'd just be getting the hostname of a container that runs `hostname`, which isn't very helpful. Instead, you would use the `Hostname()` (Go) or `hostname()` (Python and Node.js) SDK method, which returns a domain name reachable by other containers:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="cwmeT7388mg" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="YGkCbitVTYY" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="uCP-rb3rLeK" />
+
+</TabItem>
+</Tabs>
+
+In practice, you are more likely to use aliases with service bindings or endpoints, which are covered in the next section.
+
+## Exposing ports
+
+Dagger offers two methods to work with service ports:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+- Use the `WithExposedPort()` method to set ports that the container will listen on. Dagger checks the health of each exposed port prior to running any clients that use the service, so that clients don't have to implement their own polling logic.
+- Use the `Endpoint()` method to create a string address to a service's port. You can either specify a port or let Dagger pick the first exposed port.
+
+Here's an example:
+
+<Embed id="kDBfYoh2uau" />
+
+</TabItem>
+
+<TabItem value="Python">
```

Comment:
@helderco just curious but I noticed that the endpoint URL generated for the Python snippet is different from the ones in the Go and Node snippets (Go and Node produce the same output_. Do you know why?

Reply:
I ran all examples and their output was similar to Go. Checking...

Reply:
@vikram-dagger, there's a missing comma:

```diff
-                client.directory().with_new_file("index.html", "Hello world!"),
+                client.directory().with_new_file("index.html", "Hello, world!"),
```

Reply:
Modified